### PR TITLE
Movies filter - basic task solution

### PR DESCRIPTION
**Pull request by VinnieJ-2k20**
- [DEMO LINK](https://VinnieJ-2k20.github.io/react_movies-list-filter/)

### CODE REVIEW

**Review by AlieksieienkoAndriy — APPROVED** [review]

File: src/App.js [hunk 1]
```diff
@@ -2,32 +2,34 @@ import React, { Component } from 'react';
...
+      filteredMovies: moviesFromServer,
+    };
+    this.applyFilteredList = this.applyFilteredList.bind(this);
+  }
+
+  applyFilteredList(filteredList) {
+    this.setState({
+      filteredMovies: filteredList,
+    });
+  }
```
Comment (varseniuk) [inline src/App.js:20]: Using constructor and 'bind' you making the solution overcomplicated.
Let's make it more simple. 
You should filter the movies according to the 'query' and then transfer filtered movies to the 'MoviesList' component. The state should be on the top level to give us the possibility to filter films by this query. 
That means that if you want to make a separate component for the 'search input', you need somehow to transfer input value from the child component('MovieFilter') to the parent component, where you will handle that input value for filtering movies. 
Understanding the topic 'lifting state up' will help you a lot.
Long story short: your 'Filter' component should receive a 'query' and 'queryChangeHandler' from the parent. Then this component should handle input changes and transfer them to the top level using lifting state up. There you will be filtering your movies according to the new query.
If you have some questions - write to your channel or visit the q&a. I will explain

File: src/components/MovieFilter/MovieFilter.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+              type="text"
+              id="search-query"
+              value={query}
+              onChange={this.handleChange}
+              className="input"
+              placeholder="Type search word"
+            />
+          </div>
+        </div>
+      </div>
```
Comment (varseniuk) [inline src/components/MovieFilter/MovieFilter.js:65]: logic is correct, but let's change the structure using lifting state up (without constructor)

**Review by varseniuk — CHANGES_REQUESTED** [review]

File: src/App.js (hunk 1 above)
Comment (VinnieJ-2k20) [reply src/App.js:20]: @varseniuk  Okay, I`ll return to this task after learning about lifting state up.

File: src/App.js [hunk 3]
```diff
@@ -1,33 +1,35 @@
...
+import { MovieFilter } from './components/MovieFilter';
+import './App.scss';
 
 export class App extends Component {
-  state = {};
+  state = {
+    movies: moviesFromServer,
+  }
+
+  queryChangeHandler = (query) => {
```
Comment (aliceblunt8) [inline src/App.js:12]: let's separate saving query to the state and filtering movies

File: src/App.js [hunk 4]
```diff
@@ -1,33 +1,35 @@
...
+    const queryLowerCase = query.toLowerCase();
+
+    this.setState({
+      movies: moviesFromServer.filter(({ title, description }) => (
+        title.toLowerCase().includes(queryLowerCase)
+          || description.toLowerCase().includes(queryLowerCase)
+      )),
+    });
```
Comment (aliceblunt8) [inline src/App.js:20]: actually where's no need to save movies to the state and then update them, we can filter moviesFromServer
let's create a variable filteredMovies and move this code there, then we can pass them directly to the MovieList like this
`<MoviesList movies={filteredMovies} />`

**Review by aliceblunt8 — CHANGES_REQUESTED** [review]

File: src/App.js [hunk 5]
```diff
@@ -1,33 +1,37 @@
...
-  state = {};
+  state = {
+    query: '',
+  }
+
+  queryChangeHandler = (query) => {
+    this.setState({
+      query,
+    });
+  }
```
Comment (aliceblunt8) [inline src/App.js:16]: save value in lower case

File: src/components/MovieFilter/MovieFilter.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import PropTypes from 'prop-types';
+
+export class MovieFilter extends Component {
+  static propTypes = {
+    onChange: PropTypes.func.isRequired,
+  }
+
+  state = {
+    query: '',
+  }
```
Comment (aliceblunt8) [inline src/components/MovieFilter/MovieFilter.js:11]: you should have query in the state of ONE component (App), pass it here as prop

**Review by aliceblunt8 — CHANGES_REQUESTED** [review]

File: src/App.js (hunk 5 above)
Comment (VinnieJ-2k20) [reply src/App.js]: Fixed, but this means that all text in filter form will be in lower case. I haven't seen any search behave this way.

**Review by aliceblunt8 — APPROVED** [review]
